### PR TITLE
Support iOS universal links route deep linking

### PR DESCRIPTION
**Pull request by jmagman**
Support deep linking/route pushing via [universal links](https://developer.apple.com/library/archive/documentation/General/Conceptual/AppSearch/UniversalLinks.html), which are registered `http` and `https` scheme URLs, in addition to the [already-supported custom schemes](https://flutter.dev/docs/development/ui/navigation/deep-linking#enable-deep-linking-on-ios).

If the URL isn't handled by the life cycle delegate, and isn't from web browsing, send `pushRoute` to the app.
Share the open link/routing logic between `-application:continueUserActivity:restorationHandler:` and `-application:openURL:options:`.

Fixes https://github.com/flutter/flutter/issues/82550

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm [hunk 1]
```diff
@@ -141,13 +141,8 @@ static BOOL IsDeepLinkingEnabled(NSDictionary* infoDictionary) {
   }
 }
 
-- (BOOL)application:(UIApplication*)application
-            openURL:(NSURL*)url
-            options:(NSDictionary<UIApplicationOpenURLOptionsKey, id>*)options
-    infoPlistGetter:(NSDictionary* (^)())infoPlistGetter {
-  if ([_lifeCycleDelegate application:application openURL:url options:options]) {
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm:148]: This method is now shared between two app delegate methods.  Move the method-specific `_lifeCycleDelegate` check up one level to `-application:openURL:options:`.

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm [hunk 2]
```diff
@@ -181,12 +176,13 @@ - (BOOL)application:(UIApplication*)application
 - (BOOL)application:(UIApplication*)application
             openURL:(NSURL*)url
             options:(NSDictionary<UIApplicationOpenURLOptionsKey, id>*)options {
-  return [self application:application
-                   openURL:url
-                   options:options
-           infoPlistGetter:^NSDictionary*() {
-             return [[NSBundle mainBundle] infoDictionary];
-           }];
+  if ([_lifeCycleDelegate application:application openURL:url options:options]) {
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm:179]: (it was moved here)

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm [hunk 3]
```diff
@@ -229,9 +225,18 @@ - (BOOL)application:(UIApplication*)application
...
-                    continueUserActivity:userActivity
-                      restorationHandler:restorationHandler];
+  if ([_lifeCycleDelegate application:application
+                 continueUserActivity:userActivity
+                   restorationHandler:restorationHandler]) {
+    return YES;
+  }
+  if (userActivity.activityType == NSUserActivityTypeBrowsingWeb) {
+    return NO;
+  }
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm:235]: I'm not catching why this is here.  It sounds like `NSUserActivityTypeBrowsingWeb` is associated with universal links?  Wouldn't you want this?  I see this behavior matches what you want in your test so I assume it's true, I'm just not familiar with the API.

Maybe a few words as to why we return NO here?

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm [hunk 4]
```diff
@@ -78,15 +72,69 @@ - (void)skip_testLaunchUrlWithFragmentNoQueryParameter {
...
+  BOOL result = [appDelegate
+               application:[UIApplication sharedApplication]
+      continueUserActivity:userActivity
+        restorationHandler:^(NSArray<id<UIUserActivityRestoring>>* __nullable restorableObjects){
+        }];
+  XCTAssertFalse(result);
+  OCMReject([navigationChannel invokeMethod:OCMOCK_ANY arguments:OCMOCK_ANY]);
+}
+
+- (void)testUniversalLinkPushRoute {
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm:110]: These 2 tests are very similar, maybe it's worth considering pulling out a helper function to remove some of the boilerplate so the differences are more apparent.

**Review by gaaclarke — APPROVED** [review]

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm (hunk 3 above)
Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm:235]: Actually, this is a good question.  @AlBlanc I noticed you did you did this in https://github.com/flutter/engine/pull/25705 and meant to follow up.  Why did you avoid this activity type?

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate_Test.h [hunk 5]
```diff
@@ -7,9 +7,4 @@
 @interface FlutterAppDelegate (Test)
 @property(nonatomic, copy) FlutterViewController* (^rootFlutterViewControllerGetter)(void);
 
-- (BOOL)application:(UIApplication*)application
-            openURL:(NSURL*)url
-            options:(NSDictionary<UIApplicationOpenURLOptionsKey, id>*)options
-    infoPlistGetter:(NSDictionary* (^)())infoPlistGetter;
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate_Test.h:13]: I wound up getting rid of this test injection helper and mocking out `-[NSBundle objectForInfoDictionaryKey:]` instead.

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm (hunk 4 above)
Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm:142]: I pulled the boilderplate in all these tests up into `setUp`.  Hopefully it's easier to read now.

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm [hunk 6]
```diff
@@ -14,84 +14,145 @@
...
 
 @interface FlutterAppDelegateTest : XCTestCase
+@property(strong) FlutterAppDelegate* appDelegate;
+
+@property(strong) id mockMainBundle;
+@property(strong) id mockNavigationChannel;
+
+// Retain callback until the tests are done.
+// https://github.com/flutter/flutter/issues/74267
+@property(strong) id mockEngineFirstFrameCallback;
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm:24]: I only just landed https://github.com/flutter/engine/pull/27854 but this is probably a better name than `blockNoInvoker`. Open to suggestions though!

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm [hunk 7]
```diff
@@ -132,22 +132,11 @@ - (void)userNotificationCenter:(UNUserNotificationCenter*)center
...
-
-- (BOOL)application:(UIApplication*)application
-            openURL:(NSURL*)url
-            options:(NSDictionary<UIApplicationOpenURLOptionsKey, id>*)options
-    infoPlistGetter:(NSDictionary* (^)())infoPlistGetter {
-  if ([_lifeCycleDelegate application:application openURL:url options:options]) {
-    return YES;
-  } else if (!IsDeepLinkingEnabled(infoPlistGetter())) {
+- (BOOL)openURL:(NSURL*)url {
+  NSNumber* isDeepLinkingEnabled =
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm:136]: I was able to get rid of the static function here.  `-[NSBundle objectForInfoDictionaryKey:]` is now mocked in the tests.

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm [hunk 8]
```diff
@@ -14,84 +14,145 @@
...
+      [self.appDelegate application:[UIApplication sharedApplication]
+                            openURL:[NSURL URLWithString:@"http://myApp/custom/route?query=test"]
+                            options:@{}];
   XCTAssertTrue(result);
-  OCMVerify([navigationChannel invokeMethod:@"pushRoute" arguments:@"/custom/route?query=test"]);
+  OCMVerify([self.mockNavigationChannel invokeMethod:@"pushRoute"
+                                           arguments:@"/custom/route?query=test"]);
+}
+
+- (void)testLaunchUrlWithDeepLinkingNotSet {
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm:74]: New test.

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm [hunk 9]
```diff
@@ -14,84 +14,145 @@
...
+
+  BOOL result =
+      [self.appDelegate application:[UIApplication sharedApplication]
+                            openURL:[NSURL URLWithString:@"http://myApp/custom/route?query=test"]
+                            options:@{}];
+  XCTAssertFalse(result);
+  OCMReject([self.mockNavigationChannel invokeMethod:OCMOCK_ANY arguments:OCMOCK_ANY]);
+}
+
+- (void)testLaunchUrlWithDeepLinkingDisabled {
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterAppDelegateTest.mm:86]: New test.

File: shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm (hunk 3 above)
Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: I don't see a reason to not process `NSUserActivityTypeBrowsingWeb` so I removed this and the test.
https://developer.apple.com/documentation/xcode/supporting-universal-links-in-your-app

We can always add it back later if there's some good reason to.

Comment (gaaclarke) [reply shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: That sounds fine with me, we could wait a bit more to give @alblanc 24 hours to respond.  He's probably celebrating the France basketball teams win =)

Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: I was thinking I can merge this now and put the additional PR if should be excluded.  Based on my reading of https://developer.apple.com/documentation/xcode/supporting-universal-links-in-your-app it shouldn't be.
But yeah I can wait too, it's not a rush...

Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm]: I'm going to merge this. I'll change it if we hear we need to exclude `NSUserActivityTypeBrowsingWeb` for some reason.

**Comment by jmagman** [comment]
> Hi @jmagman Could you clarify what version of flutter includes this fix?

2.6.0-0.0.pre

**Comment by vasilich6107** [comment]
Hi @jmagman 
I installed 2.6.0 as you recommended.

The deep links works fine when I tap the link from email app.

The issue is if I attach the link to push notification it still opens the webview.

Is this ok for current implementation?

**Comment by janjoosse** [comment]
@vasilich6107 What setup and configuration did you do to get the deep links working (for links from an email)?

**Comment by ikurek** [comment]
@vasilich6107 Did you open the issue? I'm also having problems with deep links when they are in foreground, after deep link is fired (either from push notification or CLI) `RouteInformationParser` and `RouteInformationProvider` do not receive any kind of information about route change (although they do on Android). May be the same thing as in your case

**Comment by dragouf** [comment]
I'm using flutter 2.10.1 and when I request:
/usr/bin/xcrun simctl openurl booted "https://mydomain.com/mypath" 

it open the app but the view is not push based on the path as expected (it works perfectly on android). 

Is this still working in 2.10 ? FlutterDeepLinkingEnabled is set to true for ios. It's what I need to get the view automatically pushed right ? I use onGenerateRoute for named routes.

**Comment by percula** [comment]
@dragouf Are the instructions at https://docs.flutter.dev/development/ui/navigation/deep-linking#enable-deep-linking-on-ios still valid for universal links? I don't see why we need the custom scheme. Do we still need the <key>CFBundleURLTypes</key> entry for universal links?

**Comment by dragouf** [comment]
@percula what they call deep links is not universal links. Deep links use custom scheme when universal links (or app links for android) use https. 

If you want to use https instead of custom scheme follow the readme of this library to activate it: https://pub.dev/packages/uni_links

but do not install the library if you want the view to be pushed automatically based on the link path (the library seems to conflicts with this mechanism).

so after I enabled universal links I use named routes with something like:

`MaterialApp(onGenerateRoute: AppRouter.onGenerateRoute)`

and 

```
class AppRouter {
  static Route<dynamic> onGenerateRoute(RouteSettings settings) {
    if (settings.name != null && settings.name!.startsWith('/thepath/') {
      return CupertinoPageRoute(
        fullscreenDialog: true,
        builder: (_) => const MyWidgetView(),
      );
    }
}
```

settings.name will contain the path from your link.

**Pull request merged**